File: NeetCode150/1_valid_parenthesis.cc
// Time: O(n), n = length of string
// Space: O(1)
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> map = {{'}', '{'}, {']', '['}, {')', '('}};

        stack<char> bracketStack;

        for (const char& ch : s) {
            if (map.contains(ch) && !bracketStack.empty() &&
                map[ch] == bracketStack.top()) {
                bracketStack.pop();
            } else {
                bracketStack.push(ch);
            }
        }

        return bracketStack.empty();
    }
};

File: NeetcodeBlind75/stack/1_valid_parenthesis.py
class Solution:
    def isValid(self, s: str) -> bool:
        """
        Algorithm intuition:
            - create a hash map of opening bracket and closing bracket
            - push all the opening brackets to stack
            - if current is closing bracket and corresponding value is not closing bracket of same type return false, else pop
            - in the end if stack is empty return true else false
        Time: O(n)
        Space: O(1)
        """
        brackets = {")": "(", "]": "[", "}": "{"}

        stack = []

        for br in s:
            if br in brackets:
                if stack and stack[-1] == brackets[br]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(br)

        return True if not stack else False
